pandas/_libs/lib.pyx

```diff
@@ -920,19 +921,20 @@ def generate_slices(const intp_t[:] labels, Py_ssize_t ngroups):
 
     start = 0
     group_size = 0
-    for i in range(n):
-        lab = labels[i]
-        if lab < 0:
-            start += 1
-        else:
-            group_size += 1
-            if i == n - 1 or lab != labels[i + 1]:
-                starts[lab] = start
-                ends[lab] = start + group_size
-                start += group_size
-                group_size = 0
-
-    return starts, ends
+    with nogil:
+        for i in range(n):
+            lab = labels[i]
+            if lab < 0:
+                start += 1
+            else:
+                group_size += 1
+                if i == n - 1 or lab != labels[i + 1]:
+                    starts[lab] = start
+                    ends[lab] = start + group_size
+                    start += group_size
+                    group_size = 0
+
+    return np.asarray(starts), np.asarray(ends)
```

Comment:
FYI there is a tiny perf bump from using `starts.base` instead of `np.asarray(starts)` in these cases.

Reply:
Ah thanks, good to know